lightningd/channel_gossip.c

```diff
@@ -523,41 +765,146 @@ static void send_channel_announcement(struct channel *channel)
+					 channel->dbid,
+					 &cg->remote_sigs->node_sig,
+					 &cg->remote_sigs->bitcoin_sig);
+		return;
+
+	case CGOSSIP_ANNOUNCED:
+		/* BOLT #7:
+		 * A recipient node:
+		 *...
+		 *   - If it has sent AND received a valid `announcement_signatures`
+		 *     message:
+		 *     - If the funding transaction has at least 6 confirmations:
+		 *       - SHOULD queue the `channel_announcement` message for
+		 *         its peers.
+		 */
+		send_channel_announcement(channel);
+
+		/* Any private cupdate will be different from this, so will force a refresh. */
+		update_channel_update(channel, channel_should_enable(channel, true));
+		broadcast_new_gossip(channel->peer->ld, cg->cupdate, NULL, "channel update");
+
+		/* We need to refresh channel update every 13 days */
+		arm_refresh_timer(channel);
+
+		/* And maybe our first node_announcement */
+		channel_gossip_node_announce(channel->peer->ld);
 		return;
-	case CGOSSIP_NEED_PEER_SIGS:
+
+	case CGOSSIP_CHANNEL_ANNOUNCED_DYING:
+		/* Make sure update tells them it's disabled */
+		if (update_channel_update(channel, false)) {
+			/* We might have skipped over CGOSSIP_ANNOUNCED, so tell
+			 * gossipd about us now, so it doesn't complain. */
+			send_channel_announcement(channel);
+			/* And tell the world */
+			broadcast_new_gossip(channel->peer->ld, cg->cupdate, NULL,
+					     "channel update");
+		}
+		return;
+
+	case CGOSSIP_CHANNEL_ANNOUNCED_DEAD:
+		/* It's disabled, but gossipd has forgotten it, so no
+		 * broadcast */
+		update_channel_update(channel, false);
+		return;
+
+	case CGOSSIP_CHANNEL_UNANNOUNCED_DYING:
+	case CGOSSIP_CHANNEL_DEAD:
+		return;
+	}
+	fatal("Bad channel_gossip_state %u", cg->state);
+}
+
+static void update_gossip_state(struct channel *channel)
+{
+	enum channel_gossip_state newstate;
+	struct channel_gossip *cg = channel->channel_gossip;
+
+	newstate = derive_channel_state(channel);
+	if (newstate != cg->state)
+		set_gossip_state(channel, newstate);
+
+	switch (cg->state) {
+	case CGOSSIP_CHANNEL_DEAD:
+	case CGOSSIP_CHANNEL_UNANNOUNCED_DYING:
+	case CGOSSIP_CHANNEL_ANNOUNCED_DYING:
+	case CGOSSIP_CHANNEL_ANNOUNCED_DEAD:
+	case CGOSSIP_WAITING_FOR_USABLE:
+	case CGOSSIP_PRIVATE_WAITING_FOR_USABLE:
+		return;
+	case CGOSSIP_WAITING_FOR_MATCHING_PEER_SIGS:
+	case CGOSSIP_WAITING_FOR_ANNOUNCE_DEPTH:
 		send_channel_announce_sigs(channel);
-		/* We may already have remote signatures */
-		if (!apply_remote_sigs(channel))
-			return;
-		cg->state = CGOSSIP_ANNOUNCED;
 		/* fall thru */
+	case CGOSSIP_WAITING_FOR_SCID:
+	case CGOSSIP_PRIVATE:
+		/* Always try to send private cupdate: ignored if redundant */
+		send_private_cupdate(channel, false);
+		return;
 	case CGOSSIP_ANNOUNCED:
-		/* Any previous update was private, so clear. */
-		cg->cupdate = tal_free(cg->cupdate);
-		send_channel_announcement(channel);
+		/* If a channel parameter has changed, send new update */
+		if (update_channel_update(channel, channel_should_enable(channel, true)))
```

Comment:
This will generate a new unsigned channel update for each channel on every new block.  Not too heavy for large nodes?